---
title: "Assignment 9"
author: "Emma Afferton"
date: "November 5, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Exercise Set 1
```{r}
library(devtools)
install_github("genomicsclass/GSE5859Subset")
install_github("genomicsclass/tissuesGeneExpression")
library(tissuesGeneExpression)
library(GSE5859Subset)
data(GSE5859Subset)
source("https://bioconductor.org/biocLite.R")
```

#### Load the admissions data from the dagdata package (which is available from the genomicsclass repository):
```{r}
library(dagdata) 
data(admissions)
```

#### Familiarize yourself with this table:
```{r}
admissions
```

#### Let's compute the proportion of men who were accepted:
```{r}
index = which(admissions$Gender==1) 
accepted= sum(admissions$Number[index] * admissions$Percent[index]/100) 
applied = sum(admissions$Number[index]) 
accepted/applied

```

#### What is the proportion of women that were accepted? 
```{r}
index = which(admissions$Gender ==0)
accepted= sum(admissions$Number[index] * admissions$Percent[index]/100) 
applied = sum(admissions$Number[index]) 
accepted/applied
```

30.33% of women were accepted.

#### 2. Now that we have observed different acceptance rates between genders, test for the significance of this result. 
```{r}
index = admissions$Gender==1 
men = admissions[index,]
women = admissions[!index,] 
menYes = sum(men$Number*men$Percent/100) 
menNo = sum(men$Number*(1-men$Percent/100)) 
womenYes = sum(women$Number*women$Percent/100) 
womenNo = sum(women$Number*(1-women$Percent/100)) 
tab = matrix(c(menYes,womenYes,menNo,womenNo),2,2) 
print(chisq.test(tab)$p.val)
```

#### If you perform an independence test, what is the p-value

The p-value of 9.129492e-22. This means that we reject the null hypothesis that the gender and admissions are independent. 

#### This difference actually led to a lawsuit???. Now notice that looking at the data by major, the differences disappear.
```{r}
admissions
```

#### How can this be? This is referred to as Simpson's Paradox. In the following questions we will try to decipher why this is happening.


#### 3. We can quantify how "hard" a major is by using the percent of students that were accepted. Compute the percent that were accepted (regardless of gender) to each major and call this vector H. Which is the hardest major? 
```{r}
admissions
```

```{r}
Majors <- as.vector(unique(admissions$Major))
h <- vector("double", length(Majors))
indexh = 1
for (i in Majors) { 
  index = which(admissions$Major == i)
  accepted= sum(admissions$Number[index] * admissions$Percent[index]/100) 
  applied = sum(admissions$Number[index])
  h[[indexh]] <-  accepted/applied 
  indexh = indexh +1
}
h
```
```{r}
hardest_index <- which.min(h)
hardest_major <- Majors[hardest_index]
hardest_major
```

The hardest major was major F.

#### 4. What proportion is accepted for this major?

```{r}
proportion_accepted <- min(h)
proportion_accepted*100
```

Only 6.477 percent of people were accepted to major F. 

#### 5.  For men, what is the correlation between the number of applications across majors and H? 
```{r}
library(dplyr)
males <- admissions %>%
  filter (Gender == 1)
horder <- sort(h, decreasing = TRUE)
mcorrelations <- cbind(horder, males$Number)
mcorrelations
plot(mcorrelations)
cor(horder, males$Number)
```
The correlation for males between number of applicants and acceptance rate per major is 0.7647567. When plotted, the graph displays this positive correlation. 

#### 6.  For women, what is the correlation between the number of applications across majors and H?
```{r}
females <- admissions %>%
  filter (Gender == 0)

horder <- sort(h, decreasing = TRUE)
fcorrelations <- cbind(horder, females$Number)
fcorrelations
plot(fcorrelations)
cor(horder, females$Number)
```

The correlation for females between the number of applicants and percent acceptance for each major is -0.6743393.When plotted, the graph displays this negative correlation.

#### 7. Given the answers to the above, which best explains the differences in admission percent ages when we combine majors?

##### . A) We made a coding mistake when computing the overall admissions percentages. 
##### . B)There were more total number of women applications which made the denominator much bigger. 

##### . C) There is confounding between gender and preference for "hard" majors: females are more likely to apply to harder majors. 

##### . D)The sample size for the individual majors was not large enough to draw the correct conclusion.

For males, there is a postive correlation between the h and the number of male applications, meaning that there were a greater number of applicants for majors with higher acceptances. For females, there was a negative correlation between the number of applicants and teh acceptance rates for each majors, meaning that as the nubmer of people apply to a major increases, the acceptance rate for that major decreased. Therefore, the answer is C: women were applying to harder majors. 

### Exercise Set 2

#### We will use the Bioconductor package Biobase which you can install with install_bioc function from rafalib: Load the data for this gene expression dataset:
```{r}
install_bioc("Biobase")
install_github("genomicsclass/GSE5859")
library(tissuesGeneExpression)
library(GSE5859) 
data(GSE5859)
```

#### This is the original dataset from which we selected the subset used in GSE5859Subset. We can extract the gene expression data and sample information table using the Bioconductor functions exprs and pData like this:
```{r}
geneExpression = exprs(e) 
sampleInfo = pData(e)
```

#### 1.  Familiarize yourself with the sampleInfo table. Note that some samples were processed at different times. This is an extraneous variable and should not affect the values in geneExpression. However, as we have seen in previous analyses, it does appear to have an effect so we will explore this here. You can extract the year from each date like this:
```{r}
year = format(sampleInfo$date,"%y")
```

#### Note that ethnic group and year is almost perfectly confounded:
```{r}
table(year,sampleInfo$ethnicity)
```

#### 2.  For how many of these years do we have more than one ethnicity represented? 
```{r}
yr_eth <- table(year,sampleInfo$ethnicity)
```
```{r}
numyears = 0 
yr_eth <- table(year,sampleInfo$ethnicity)
q = c()

for (i in seq_along(1:nrow(yr_eth))){
  if((yr_eth[i,1] == 0 & yr_eth[i,2] == 0) | (yr_eth[i,1] ==0 & yr_eth[i,3] ==0) | (yr_eth[i,2] == 0 & yr_eth[i,3] == 0)){
    numyears = numyears
  }
  else{

    numyears = numyears +1
  }
}

numyears
```

There are 2 years with more than one ethnicity represented.  

#### 3. Repeat the above exercise, but now,instead of year,consider the month as well.Specifically, instead of the year variable defined above use:

```{r}
month.year = format(sampleInfo$date,"%m%y")
```

#### For what proportion of these month.year values do we have more than one ethnicity represented? 
```{r}
ym_eth <- table(month.year,sampleInfo$ethnicity)
num_ym = 0
for (i in seq_along(1:nrow(ym_eth))){
  if((ym_eth[i,1] == 0 & ym_eth[i,2] == 0) | (ym_eth[i,1] ==0 & ym_eth[i,3] ==0) | (ym_eth[i,2] == 0 & ym_eth[i,3] == 0)){
    num_ym = num_ym
  }
  else{
    num_ym = num_ym +1
  }
}
num_ym
prop_multiple_eth <- ((nrow(ym_eth) - num_ym)/nrow(ym_eth))*100
prop_multiple_eth 
ym_eth
```

95.2381% of the year-month combos have only one ethnicity represented. 

#### 4. Perform a t-test (use rowttests) comparing CEU samples processed in 2002 to those processed in 2003. Then use the qvalue package to obtain q-values for each gene. How many genes have q-values < 0.05 ?
```{r}
library(dplyr)
install_bioc("genefilter")
library("genefilter")
```

```{r}
dates <- pData(e)$date 

index = which(sampleInfo$ethnicity == "CEU")
year = factor(format(dates, "%y"))[index]
year = droplevels(year, c("04","05","06"))
pvals<- rowttests(geneExpression[,index],year)$p.value
```
```{r}
biocLite("qvalue")
library(qvalue)
```
```{r}
qvals<-qvalue(pvals)$qvalue
index <- which(qvals <0.05)
result <- res$p.value[index]
length(result)
```

There aere 4308 genes with a q-value less than 0.05

#### 5. What is the estimate of pi0 provided by qvalue:
```{r}
pi0<-qvalue(pvals)$pi0
pi0
```

The pi0 is 0.3628642.

#### 6. Now perform a t-test (use rowttests) comparing CEU samples processed in 2003 to those processed in 2004.Then use the qvalue package to obtain q-values for each gene.How many genes have q-values less than 0.05? 

```{r}
index = which(sampleInfo$ethnicity == "CEU")
year = factor(format(dates, "%y"))[index]
year = droplevels(year, c("02","05","06"))
pvals<- rowttests(geneExpression[,index],year)$p.value

qvals<-qvalue(pvals)$qvalue
index <- which(qvals <0.05)
result0 <- res$p.value[index]
length(result)
```

There are 4308 genes with q-values less than 0.05.

#### 7. Now we are going to compare ethnicities as was done in the original publication in which these data were first presented. Use the qvalue function to compare the ASN population to the CEU population. Once again, use the qvalue function to obtain q-values. How many genes have q-values < 0.05 ? 

```{r}
index = which(sampleInfo$ethnicity==c("ASN", "CEU"))
ethnicity = factor(sampleInfo$ethnicity)[index]
ethnicity = droplevels(ethnicity, "HAN")
pvals<- rowttests(geneExpression[,index],ethnicity)$p.value

qvals<-qvalue(pvals)$qvalue
index <- which(qvals <0.05)
result <- res$p.value[index]
length(result)
```

There are 6356 genes with q-values less than 0.05.

#### 8. Over 80% of genes are called differentially expressed between ethnic groups. However, due to the confounding with processing date, we need to confirm these differences are actually due to ethnicity. This will not be easy due to the almost perfect confounding. However, above we noted that two groups were represented in 2005. Just like we stratified by majors to remove the "major effect" in our admissions example, here we can stratify by year and perform a t-test comparing ASN and CEU, but only for samples processed in 2005. How many genes have q-values < 0.05 

```{r}
year = format(sampleInfo$date,"%y")
sampleinfoyear <-cbind(year, sampleInfo)


index = which(sampleinfoyear$year=="05")
ethnicity = factor(sampleinfoyear$ethnicity)[index]
ethnicity = droplevels(ethnicity, "HAN")
pvals<- rowttests(geneExpression[,index],ethnicity)$p.value

qvals<-qvalue(pvals)$qvalue
index <- which(qvals <0.05)
result <- res$p.value[index]
length(result)
```

There are 560 genes with q-values less than 0.05.

#### Notice the dramatic drop in the number of genes with q-value < 0.05 when we fix the year. However, the sample size is much smaller in this latest analysis which means we have less power:
```{r}
table(sampleInfo$ethnicity[index])
```

#### 9. To provide a more balanced comparison, we repeat the analysis, but now taking 3 random CEU samples from 2002. Repeat the analysis above, but comparing the ASN from 2005 to three random CEU samples from 2002. Set the seed at 3, set.seed(3) How many genes have q-values < 0.05 ?

```{r}
year = format(sampleInfo$date,"%y")
sampleinfoyear <-cbind(year, sampleInfo)


index = which(sampleinfoyear$year=="05")
ethnicity = factor(sampleinfoyear$ethnicity)[index]
ethnicity = droplevels(ethnicity, "HAN", "CEU")
pvals<- rowttests(geneExpression[,index],ethnicity)$p.value

qvals<-qvalue(pvals)$qvalue
index <- which(qvals <0.05)
result <- res$p.value[index]
length(result)
```
```{}
set.seed(3)
index = which((sampleinfoyear$year=="02" & sampleinfoyear$ethnicity== "ASN") & (sampleinfoyear$year == "05" & sampleinfoyear$ethnicity = "CEU"))

ethnicity = factor(sampleinfoyear$ethnicity)[index]
ethnicity = droplevels(ethnicity== "HAN")
pvals<- rowttests(geneExpression[,index],ethnicity)$p.value

qvals<-qvalue(pvals)$qvalue
index <- which(qvals <0.05)
result <- res$p.value[index]
length(result)

```


### Exercise Set 3

#### For the dataset we have been working with, models do not help due to the almost perfect confounding. This is one reason we created the subset dataset:
```{r}
library(GSE5859Subset) 
data(GSE5859Subset)
```
#### Here we purposely confounded month and group (sex), but not completely:

```{r}
sex = sampleInfo$group 
month = factor( format(sampleInfo$date,"%m")) 
table( sampleInfo$group, month)
```

#### 1. Using the functions rowttests and qvaluec ompare the two groups.Because this is a smaller dataset which decreases our power, we will use the more lenient FDR cut-off of 10%. How many gene have q-values less than 0.1? 
```{r}
library(GSE5859Subset) 
data(GSE5859Subset)
```

```{r}
res <- rowttests(geneExpression,as.factor( sampleInfo$group )) 

qvals <-qvalue(res$p.value)$qvalue
index <- which(qvals <0.1)

length((res$p.value[index]))
```

There are 59 genes with q-values less than 0.1

### 2. Note that sampleInfo$group here presents males and females. Thus, we expect differences to be in on chrY and, for genes that escape inactivation, chrX. We do not expect many autosomal genes to be different between males and females.This gives us an opportunity to evaluate false and true positives with experimental data. For example, we evaluate results using the proportion genes of the list that are on chrX or chrY. For the list calculated above, what proportion of this list is on chrX or chrY? 

```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("hgfocus.db", version = "3.8")
```
```{r}
library(hgfocus.db) ##install from Bioconductor 

annot <- select(hgfocus.db, keys=featureNames(e), keytype="PROBEID", columns=c("CHR")) ##for genes with multiples, pick one 
annot <-annot[match(featureNames(e),annot$PROBEID),] 
annot$CHR <- ifelse(is.na(annot$CHR),NA,paste0("chr",annot$CHR))
head(annot)

```
```{r}
res <- rowttests(geneExpression,as.factor( sampleInfo$group ))
qvals <- qvalue(res$p.value)$qvalue 
index <- which(qvals<0.1)
list<- length((res$p.value[index]))
chr <- annot$CHR

y_sum <- sum(chr[index]=="chrY",na.rm=TRUE)
x_sum <- sum(chr[index]=="chrX",na.rm=TRUE)
proportion <- (y_sum + x_sum)/list
proportion *100

```

33.89% of the genes represented are in either chromosome X or Y

#### 3. We can also check how many of the chromosomes X and Y genes we detected  as different. How many are on Y? 
```{r}
y_sum <- sum(chr[index]=="chrY",na.rm=TRUE)
y_sum
```

There are 8 genes on chromosome Y.

#### 4.  Now for the autosomal genes (not on chrX and chrY) for which q-value < 0.1, perform a t-test comparing samples processed in June to those processed in October. What proportion of these have p-values <0.05 ?
```{r}
index2 <- index[(!chr[index] == "chrY") &!chr[index] == "chrX"]

pvals<- rowttests(geneExpression[index2,],month)$p.value

qvals<-qvalue(pvals, lambda = 0)$qvalue
index <- which(qvals <0.05)
result <- res$p.value[index]
length(result)
length(index2)
propor <- length(result)/length(index2)
propor*100
```

87.2% o the autosomla genes are significant. 

#### 5. The above result shows that the great majority of the autosomal genes show differences due to processing data. This provides further evidence that confounding is resulting in false positives. So we are going to try to model the month effect to better estimate the sex effect. We are going to use a linear model: Which of the following creates the appropriate design matrix? 

##### . A) X = model.matrix(???sex+ethnicity) 
##### . B) X = cbind(sex,as.numeric(month)) 
##### . C) It can't be done with one line. 
##### . D) X = model.matrix(???sex+month)


To create a linear model to model month as having an eaffect with sex, it should be modelled as D: X = model.matrix(???sex+month).


#### 6. Now use the X defined above,to fit a regression model using lm for each gene.You can obtain p-values for estimated parameters using summary. Here is an example
````{r}
X = model.matrix(~sex+month) 
i = 234 
y = geneExpression[i,] 
fit = lm(y~X) 
summary(fit)$coef
```
```{r}

X = model.matrix(~sex+month)

res <- t( sapply(1:nrow(geneExpression),function(j){ 
  y <- geneExpression[j,] 
  fit = lm(y~X)
  summary(fit)$coef[2,c(1,4)]
}))

res <- data.frame(res)
names(res) <- c("dm","p.value")
qvals <- qvalue(res$p.value)$qvalue 
index <- which(qvals<0.1)
length(index)
```

#### How many of the q-values for the group comparison are now <0.1? Note the big drop from what we obtained without the correction. 

There are 17 genes with q values less than 0.1.

#### 7. With this new list, what proportion of these are chrX and chrY? Notice the big improvement. 
```{r}
chrx<- sum(chr[index]=="chrY",na.rm=TRUE)
chry<- sum(chr[index]=="chrX",na.rm=TRUE)
propor <- sum((chrx + chry)/ length(index))
propor*100
```

88.24% of the genes are on chromosome x or y

#### 8. How many on Y or X?
```{r}
chry
chrx
```

There are 9 genes on chromosome y and 6 on chromosome x.

#### 9. Now from the linear model above,extract the p-values related to the coefficient representing the October versus June differences using the same linear model. How many of the q-values for the month comparison are now <0.1? This approach is basically the approach implemented by Combat.

```{r}
library(sva)
```
```{r}
mod <- model.matrix(~sex)
cleandat <- ComBat(geneExpression, month, mod)
res<-genefilter::rowttests(cleandat,factor(sex))
qvals <- qvalue(res$p.value)$qvalue 
index <- which(qvals<0.1) 
length(index)
```
```{r}
X = model.matrix(~sex+month) 
i = 234 
y = geneExpression[i,] 
fit = lm(y~X) 
summary(fit)$coef
```
```{r}
X = model.matrix(~sex+month)

res <- t( sapply(1:nrow(geneExpression),function(j){ 
  y <- geneExpression[j,] 
  fit = lm(y~X)
  summary(fit)$coef[3,c(1,4)]
}))

res <- data.frame(res)
names(res) <- c("dm","p.value")
qvals <- qvalue(res$p.value)$qvalue 
index <- which(qvals<0.1)
length(index)
```


### Exercise Set 4

#### We will continue to use this dataset:
```{r}
library(Biobase) 
library(GSE5859Subset) 
data(GSE5859Subset)
```

#### 1. Suppose you want to make an MAplot of the first two samples y = geneExpression[,1:2]. Which of the following projections gives us the projection of y so that column2 versus column 1 is an MA plot? 

```{r}
y = geneExpression[,1:2]
head(y)
```

##### A. y(1/???2 1/???2 1???2 ???1/???2)

##### B. y(1 1 1 ???1) 

##### C. (1 1 1 ???1)y 

##### D. (1 1 1 ???1)y??? 

I think that the projection of y described in C. would create the MA plot of column 2 vs. column 1. 


#### 2. Say Y is M ×N, in the SVD Y = UDV??? which of the following is not correct? 

##### . A) DV??? are the new coordinates for the projection U???Y 

##### . B) UD are the new coordinates for the projection Y V 

##### . C) D are the coordinates of the projection U???Y 

##### . D) U???Y is a projection from an N-dimensional to M-dimensional subspace.

I dont think C is true: UD are the new coordinates for the projection of YV.

#### 3. Define:

```{r}
y = geneExpression - rowMeans(geneExpression)
```

#### Compute and plot an image of the correlation for each sample. Make two image plots of these correlations. In the first one, plot the correlation as image. In the second, order the samples by date and then plot an image of the correlation.Theonlydifferenceintheseplots is the order in which the samples are plotted. Based on these plots, which of the following you would say is true? 
```{r}
library(RColorBrewer)
library(rafalib)
n <- nrow(pData(e)) 
o <- order(pData(e)$date)

Y=exprs(e)
cors=cor(Y-rowMeans(Y))
cols=colorRampPalette(rev(brewer.pal(11,"RdBu")))(100)
mypar()
image(1:n,1:n,cors,xaxt="n",yaxt="n",col=cols,xlab="",ylab="",zlim=c(-1,1))

Y=exprs(e)[,o] 
cors=cor(Y-rowMeans(Y))
cols=colorRampPalette(rev(brewer.pal(11,"RdBu")))(100)
mypar()
image(1:n,1:n,cors,xaxt="n",yaxt="n",col=cols,xlab="",ylab="",zlim=c(-1,1))
       
```

##### . A) The samples appear to be completely independent of each other. 

##### . B) Sex seems to be creating structures as evidenced by the two cluster of highly correlated samples. 

##### . C) There appear to be only two factors completely driven by month. 

##### . D) The fact that in the plot ordered by month we see two groups mainly driven by month,and within these we see subgroups driven by date, seems to suggest date more than month per se are the hidden factors.

The correlations plots look similar when your order by date versus when you do not order by date. This suggests that even when not ordered, date has an effect in the correlation. Therefore, I would guess that b. is true, that sex seems to be creating structures which are higly correlated. 

#### 4.  Based on the correlation plots above, we could argue that there are at least two hidden factors. Using PCA estimate these two factors. Specifically, apply the svd to y and use the first two PCs as estimates. Which command gives us these estimates? 

##### . A) pcs = svd(y)$v[1:2,] 

##### . B) pcs = svd(y)$v[,1:2] 

##### . C) pcs = svd(y)$u[,1:2] 

##### . D) pcs = svd(y)$d[1:2]

B will should give the pcs. A previous code was run in the chapter that looked at the first two facors. The s$v is needed to get these factors:

```{}
s <- svd(Y) 
What <- t(s$v[,1:2]) 
colnames(What)<-colnames(Y) 
round(What,2)
```

#### 5.  Plot each of the estimated factors ordered by date.Use color to denote month.The first factor is clearly related to date. Which of the following appear to be most different according to this factor? 

```{r}
times <-sampleInfo$date 

s <- svd(y) 
mypar(1,1) 
o<-order(times) 
cols <- as.numeric(month) 
plot(s$v[o,],pch=21,cex=1.25,bg=cols[o],ylab=" PC") 

plot(s$v[o,1],pch=21,cex=1.25,bg=cols[o],ylab="First PC") 

```
```{r}
times
```
##### . A) June 23 and June 27 

##### . B) Oct 07 and Oct 28 

##### . C) June 10 and June 23 

##### . D) June 15 and June 24 

The dates that seems to be the most different look like the first couple od dates, so I would guess c) June 10 and June 23.

#### 6. Use the svd function to obtain the principal components (PCs) for our detrended gene expression data y. How many PCs explain more than 10% of the variability?

```{r}
y = geneExpression - rowMeans(geneExpression)
s <- svd(y)
pcs = svd(y)$v
mypar(1,1) 
plot(s$d^2/sum(s$d^2),ylab="% variance explained",xlab="Principal component")

abline(h = 0.1)

```

There are two PCs that explain for than 10% of the variability. 

#### 7.  Which PC most correlates (negative or positive correlation) with month?
```{r}
corr <- sapply(1:ncol(s$v),function(i){ 
  fit <- lm(s$v[,i]~as.factor(month)) 
  return( summary(fit)$adj.r.squared ) }) 
mypar() 
plot(seq(along=corr), corr, xlab="PC")

```

The first PC shows the highest correlation

#### 8. What is this correlation (in absolute value)? 

```{r}
corr[1]
```

The correlation of the first PC is 0.6743

#### 9.  Which PC most correlates (negative or positive correlation) with sex?

```{r}
corr <- sapply(1:ncol(s$v),function(i){ 
  fit <- lm(s$v[,i]~as.factor(sex)) 
  return( summary(fit)$adj.r.squared ) }) 
mypar() 
plot(seq(along=corr), abs(corr), xlab="PC")
```

It looks like the first PC has the greatest correlation. 

#### 10. What is this correlation (in absolute value)? 

```{r}
corr[1]
```

The correlation is 0.3612105

#### 11. Now instead of using month, which we have shown does not quite describe the batch, add the two estimated factors s$v[,1:2] to the linear model we used above. Apply this model to each gene and compute q-values for the sex difference. How many q-values< 0.1 for the sex comparison? 

```{r}
pcs = svd(y)$v[,1:2] 
X = model.matrix(~sex+pcs)

res <- t( sapply(1:nrow(geneExpression),function(j){ 
  y <- geneExpression[j,] 
  fit = lm(y~X)
  summary(fit)$coef[2,c(1,4)]
}))

res <- data.frame(res)
names(res) <- c("dm","p.value")
qvals <- qvalue(res$p.value)$qvalue 
index <- which(qvals<0.1)
length(index)
```

There are 14 genes with q < 0.1

#### 12. What proportion of the genes are on chromosomes X and Y?

```{r}
chrx<- sum(chr[index]=="chrY",na.rm=TRUE)
chry<- sum(chr[index]=="chrX",na.rm=TRUE)
propor <- sum((chrx + chry)/ length(index))
propor*100
```

100% of the chromosomes are on X and Y! 

### Exercise Set 5

#### In this section we will use the sva function in the sva package (available from Bioconductor) and apply it to the following data:

```{r}
library(sva) 
library(Biobase)
library(GSE5859Subset) 
data(GSE5859Subset)
```

#### 1.  In a previous section we estimated factors using PCA, but we noted that the first factor was correlated with our outcome of interest:

```{r}
s <- svd(geneExpression-rowMeans(geneExpression)) 
cor(sampleInfo$group,s$v[,1])
```

#### The svafit function estimates factors, but downweighs the genes that appear to correlate with the outcome of interest. It also tries to estimate the number of factors and returns the estimated factors like this:

```{r}
sex = sampleInfo$group 
mod = model.matrix(~sex) 
svafit = sva(geneExpression,mod) 
head(svafit$sv)
```

#### The resulting estimated factors are not that different from the PCs.

```{r}
for(i in 1:ncol(svafit$sv)){ 
  print( cor(s$v[,i],svafit$sv[,i]) ) }
```

#### Now fit a linear model to each gene that instead of month includes these factors in the model. Use the qvalue function. How many genes have q-value < 0.1? 
```{r}
library (limma)


sex = sampleInfo$group 
mod = model.matrix(~sex) 
svafit = sva(geneExpression,mod)

svaX<-model.matrix(~sex+svafit$sv) 
lmfit <- lmFit(geneExpression,svaX) 
tt<- lmfit$coef[,2]*sqrt(lmfit$df.residual)/(2*lmfit$sigma)

res <- data.frame(dm= -lmfit$coef[,2], p.value=2*(1-pt(abs(tt),lmfit$df.residual[1]) ) ) 

qvals <- qvalue(res$p.value)$qvalue 
index <- which(qvals<0.1)

length(index)
```

14 genes have a q-value less than 0.1.

#### 2.How many of these genes are from chrY or chrX

```{r}
chrx<- sum(chr[index]=="chrY",na.rm=TRUE)
chry<- sum(chr[index]=="chrX",na.rm=TRUE)
chrx
chry
sum(chrx + chry)
````

13 of these genes are from either chrX or chrY.